enterprise/server/remote_execution/cgroup/cgroup.go

```
@@ -31,6 +32,105 @@ const (
+	}
+	if s.CpuUclampMin != nil {
+		m["cpu.uclamp.min"] = fmtPercent(s.GetCpuUclampMin())
+	}
+	if s.CpuUclampMax != nil {
+		m["cpu.uclamp.max"] = fmtPercent(s.GetCpuUclampMax())
+	}
+	if s.MemoryThrottleLimitBytes != nil {
+		m["memory.high"] = strconv.Itoa(int(s.GetMemoryThrottleLimitBytes()))
+	}
+	if s.MemoryLimitBytes != nil {
+		m["memory.max"] = strconv.Itoa(int(s.GetMemoryLimitBytes()))
+	}
+	if s.MemorySoftGuaranteeBytes != nil {
+		m["memory.low"] = strconv.Itoa(int(s.GetMemorySoftGuaranteeBytes()))
+	}
+	if s.MemoryMinimumBytes != nil {
+		m["memory.min"] = strconv.Itoa(int(s.GetMemoryMinimumBytes()))
+	}
+	if s.SwapThrottleLimitBytes != nil {
+		m["memory.swap.high"] = strconv.Itoa(int(s.GetSwapThrottleLimitBytes()))
+	}
+	if s.SwapLimitBytes != nil {
+		m["memory.swap.max"] = strconv.Itoa(int(s.GetSwapLimitBytes()))
+	}
+	if blockDevice != nil {
+		if s.BlockIoLatencyTargetUsec != nil {
+			m["io.latency"] = fmt.Sprintf("%d:%d target=%d", blockDevice.Maj, blockDevice.Min, s.GetBlockIoLatencyTargetUsec())
+		}
+		if s.BlockIoWeight != nil {
+			m["io.weight"] = fmt.Sprintf("%d:%d %d", blockDevice.Maj, blockDevice.Min, clampWeight(s.GetBlockIoWeight()))
+		}
+		var limitFields []string
+		limits := s.GetBlockIoLimit()
+		if limits != nil {
+			if limits.Riops != nil {
+				limitFields = append(limitFields, fmt.Sprintf("riops=%d", limits.GetRiops()))
+			}
+			if limits.Wiops != nil {
+				limitFields = append(limitFields, fmt.Sprintf("wiops=%d", limits.GetWiops()))
+			}
+			if limits.Rbps != nil {
+				limitFields = append(limitFields, fmt.Sprintf("rbps=%d", limits.GetRbps()))
+			}
+			if limits.Wbps != nil {
+				limitFields = append(limitFields, fmt.Sprintf("wbps=%d", limits.GetWbps()))
+			}
+		}
+		if len(limitFields) > 0 {
+			m["io.max"] = fmt.Sprintf("%d:%d %s", blockDevice.Maj, blockDevice.Min, strings.Join(limitFields, " "))
+		}
+	}
+	return m, nil
+}
+
+func clampWeight(v int64) int64 {
```

I'm thinking that it might be better to fail requests with invalid values instead of clamping them. That will make it faster to find bugs in the execution and scheduler servers.

---

agreed, done